### PR TITLE
refactor: keep client initialization state internal [WPB-12004]

### PR DESCRIPTION
**Pull request by SimonThormeyer**
# What's new in this PR
See above

----
##### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

**Comment by typfel** [comment]
@SimonThormeyer set the target branch on the PR to `feat/transaction` so that the diff makes sense.

File: crypto/src/mls/client/mod.rs [hunk 1]
```diff
@@ -192,12 +213,17 @@ impl Client {
...
             id: client_id.clone(),
             identities: ClientIdentities::new(stored_skp.len()),
             keypackage_lifetime: KEYPACKAGE_DEFAULT_LIFETIME,
         };
 
+        let mut inner_lock = self.state.lock().await;
+        *inner_lock = Some(client);
+        // Drop the lock because it will be acquired below when the identities are saved.
+        drop(inner_lock);
+
```
Comment (coriolinus) [inline crypto/src/mls/client/mod.rs:226]: You use this pattern several times. We can automate it a bit:

```rust
async fn replace_inner(&self, new_inner: ClientInner) {
    let mut inner_lock = self.state.lock().await;
    *inner_lock = Some(new_inner);
}
```

Which lets us simplify the usage:

```rust
self.replace_inner(ClientInner{
    id: client_id.clone(),
    identities: ClientIdentities::new(stores_skp.len()),
    keypackage_lifetime: KEYPACKAGE_DEFAULT_LIFETIME,
}).await;
```

File: crypto/src/mls/client/mod.rs [hunk 2]
```diff
@@ -454,14 +515,6 @@ impl Client {
     }
 }
 
-impl PartialEq for Client {
```
Comment (coriolinus) [inline crypto/src/mls/client/mod.rs:457]: No reason not to keep implementing `PartialEq` and `Eq` for `Client`:

```rust
let Some(ClientInner{id: sid,..}) = self.state.lock().await.deref() else { return false };
let Some(ClientInner{id: oid,..}) = other.state.lock().await.deref() else { return false };
sid == oid
```

**Review by coriolinus — COMMENTED** [review]
Not approving yet because I think swapping out the `Mutex` for a `RwLock` is actually important. The rest of the comments can be considered advisory/recommendations but not critical. But overall LGTM.

File: crypto/src/mls/client/mod.rs (hunk 2 above)
Comment (coriolinus) [reply crypto/src/mls/client/mod.rs:457]: I was wrong here: of course we can't `.await` within a non-async function, such as `PartialEq`.

**Review by coriolinus — APPROVED** [review]
Looks great! We can defer the question of replacing `async-lock` with `tokio` until later. If we happen to start using surrealdb, and if that brings in `tokio` already, then it's relatively much easier to make that switch.

**Pull request merged**